README.md
<!--
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2021-03-25 16:15:15
 * @LastEditors: zhai
 * @LastEditTime: 2021-04-16 14:04:24
-->
# seat reminder

## 久坐提醒插件

对于程序员来说，经常会遇到项目周期短的状况。白天开会讨论需求，晚上加班写代码，熬夜发版本，上线改 Bug。工作日就是：开会、写代码、倒水、上厕所、抽烟、改 Bug，一天最开心的时刻就是带薪拉屎。

好不容易到了周末，有时还会通宵玩游戏，半个月也不运动一次。再好的身体，也会被这种高强度的工作，无规律的生活所击垮。随着年龄越来越大，加上每天久坐不起来运动运动，这样下去身体真的顶不住，久坐有挺多危害的： 

0. 久坐可能会导致心脑血管疾病增加

1. 久坐可能会导致免疫力低下

1. 久坐可能会导致损脑伤胃

1. 久坐可能会得痔*

1. ...

这款插件可以每[ 1 | 30 | 45 | 60 | 90 | 120 ]分钟提醒一次该起来起来运动啦，并且展示一些骚骚的动图。12个经典小动作让你肩不痛，腰不酸，腿不麻，头不昏，眼不花！

![img](https://cdn.jsdelivr.net/gh/kanbang/resource@main/pic/seatreminder.gif)

俗话说“人怕不动，脑怕不用”，如果动起来还能锻炼大脑，何不从现在开始呢？

这个插件装备了 12 种缓解疲劳的动作，自己去解锁吧。。。

我在代码中加入了鼠标/键盘监听的功能，如果鼠标/键盘五分钟没有活动过，那就认为已经在休息了。这时候会停止计时，只有下一次鼠标/键盘活动时才会重新计时。**喜欢给个 Star！**

## 鸣谢
 This is inspired by https://github.com/s-unscrupulous/idea_seat


## Release Notes

### 0.0.5

Initial release of seat reminder


## 附：久坐的危害

0. 久坐状态下容易造成心肌萎缩、动脉硬化，甚至诱发冠心病、心肌梗塞等心脑血管疾病
1. 久坐会造成大脑供血不足，损伤大脑健康，表现为反应迟钝、疲困乏力、记忆力减退等
1. 久坐会增加患老年痴呆的风险，长时间久坐还会促进血栓的形成，堵塞脑血管，诱发中风
1. 久坐会降低细胞对胰岛素的敏感性，导致胰腺不断分泌胰岛素，影响体内糖代谢循环，容易诱发糖尿病等疾病",
1. 久坐会影响臀部的血流循环，加上臀部容易出汗，湿气排出不畅，极易引起真菌感染，使臀部皮肤过敏或长湿疹
1. 久坐会压迫男性的前列腺，使前列腺血流循环不畅，代谢产物难以排出，就会堵塞前列腺管，导致前列腺充血、淤血，诱发前列腺炎
1. 久坐状态下热量难以消耗，就会变成脂肪堆积在体内，使体重增加，引发肥胖，甚至加大患慢性病的几率



CHANGELOG.md
<!--
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2021-03-25 16:15:15
 * @LastEditors: zhai
 * @LastEditTime: 2021-04-16 13:58:48
-->

## 0.0.5 (2021-04-15)
* Initial Release

src/reminder.ts
/*
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2021-04-15 16:14:29
 * @LastEditors: zhai
 * @LastEditTime: 2021-04-15 17:45:23
 */

import * as vscode from 'vscode';
import { singleton } from "tsyringe";
import { ChildProcess } from 'child_process';
import { StatusBarItem } from 'vscode';
import { formatMinutesText } from './extension';


const MAX_IDLE = 5;

@singleton()
export class Reminder {
    private _minutespass: number = 0;
    private _minutes: number = 0;
    private _idle: number = 0;
    private _enable: boolean = false;

    constructor(private _statusItem: StatusBarItem) {

        setInterval(() => {
            // 5分钟内有操作
            if (this._enable && this._idle < MAX_IDLE) {
                this._minutespass += 1;

                this.updateStatusItem();
                
                if (this._minutespass >= this._minutes) {
                    this._minutespass = 0;

                    this.remind();
                }
            }

        }, 1000 * 60);
    }

    set minutes(val: number) {
        this._minutes = val;
        this.updateStatusItem();
    }

    set minutespass(val: number) {
        this._minutespass = val;
        this.updateStatusItem();
    }

    set enable(val: boolean) {
        this._enable = val;
    }

    remindLater(minutes: number = 5) {
        this.minutespass = this._minutes - minutes;
    }

    active() {
        this._idle = 0;
    }

    async remind() {
        const process = await vscode.commands.executeCommand<ChildProcess>('kk.seat.remind');
        console.log("remind process: ", process);
        
        this.minutespass = 0;
    }

    updateStatusItem() {
        this._statusItem.text = formatMinutesText(this._minutespass, this._minutes);
    }
}


webpack.config.js
/*
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2021-03-25 16:15:15
 * @LastEditors: zhai
 * @LastEditTime: 2021-04-02 15:40:27
 */
//@ts-check

'use strict';

const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');



/**@type {import('webpack').Configuration}*/
const config = {
  target: 'node', // vscode extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/
  mode: 'none', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')

  entry: './src/extension.ts', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
  output: {
    // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
    path: path.resolve(__dirname, 'dist'),
    filename: 'extension.js',
    libraryTarget: 'commonjs2'
  },
  devtool: 'nosources-source-map',
  externals: {
    vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
  },
  resolve: {
    // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [{
      test: /\.ts$/,
      exclude: /node_modules/,
      use: [{
        loader: 'ts-loader'
      }]
    }]
  }
};


module.exports = config;


src/extension.ts
/*
 * @Descripttion: 
 * @version: 0.x
 * @Author: zhai
 * @Date: 2021-03-25 16:15:15
 * @LastEditors: zhai
 * @LastEditTime: 2021-04-15 17:59:50
 */
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below

import * as vscode from 'vscode';

import * as path from 'path';

import "reflect-metadata";
import { container } from "tsyringe";
import { ELECTRON_ENV_VARS, ELECTRON_INSTALL_PATH, OUTPUT_CHANNEL } from "./lib/api";

import { ElectronInstaller } from "./lib/electron-installer";
import { ElectronManager } from "./lib/electron-manager";
import { ChildProcess } from 'child_process';

import * as dialog from './util/webviewDialog';
import { Reminder } from './reminder';
import { allowedNodeEnvironmentFlags } from 'node:process';

export function formatMinutesText(minutespass: number, minutes: number | string) {
	return `$(clock) [${minutespass}/${minutes}]分钟`;
}


let reminder: Reminder;

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "seat" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json

	const spawn_env = JSON.parse(JSON.stringify(process.env));
	delete spawn_env.ATOM_SHELL_INTERNAL_RUN_AS_NODE;
	delete spawn_env.ELECTRON_RUN_AS_NODE;

	container.register(ELECTRON_INSTALL_PATH, { useValue: "./electron" });
	container.register(OUTPUT_CHANNEL, { useValue: vscode.window.createOutputChannel('VS ELECTRON') });
	container.register(ELECTRON_ENV_VARS, { useValue: spawn_env });

	const installer = container.resolve(ElectronInstaller);
	const electronManager = container.resolve(ElectronManager);

	const statusItemEnable = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 1000);
	statusItemEnable.command = 'kk.seat.switch_enable';
	statusItemEnable.show();

	const statusItemMinutes = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 999);
	statusItemMinutes.command = 'kk.seat.set_minutes';

	// const reminderOnText = "$(getting-started-item-checked) 久坐提醒: 开";
	// const reminderOffText = "$(getting-started-item-unchecked) 久坐提醒: 关";

	const reminderOnText = "$(pass) 久坐提醒: 开";
	const reminderOffText = "$(circle-slash) 久坐提醒: 关";

	{
		let config = vscode.workspace.getConfiguration("seat");
		let enable = config.get<boolean>('enable', true);
		statusItemEnable.text = enable ? reminderOnText : reminderOffText;

		let minutes = config.get<number>('reminderIntervalInMinutes', 45);
		statusItemMinutes.text = formatMinutesText(0, minutes);

		if (enable) {
			statusItemMinutes.show();
		}

		reminder = new Reminder(statusItemMinutes);
		reminder.minutes = minutes;
		reminder.enable = enable;
	}

	let disposables = [
		vscode.commands.registerCommand('kk.demo.helloWorld', () => {
			vscode.window.showInformationMessage('Hello World from seat!');
		}),

		vscode.commands.registerCommand('kk.electron.validate', (version?: string) => installer.validate(version)),

		vscode.commands.registerCommand('kk.electron.install', (version?: string) => installer.install(version)),

		vscode.commands.registerCommand('kk.electron.run', async (file: string, ...args) => {
			let version = "latest";
			let params = args;

			if (args[0].hasOwnProperty("version") && args[0].hasOwnProperty("params")) {
				version = args[0].version;
				params = args[0].params;
			}
			return electronManager.run(file, version, ...params);
		}),

		vscode.commands.registerCommand('kk.seat.remind', async () => {
			let version = "v12.0.1";
			let params: string[] = [];

			const file = path.resolve(__dirname, `electron/main/index.js`);

			const child = await electronManager.run(file, version, ...params);
			if (child) {
				child.on("exit", (code) => {
					console.log('child process exit..');

				});

				child.on('message', (msg) => {
					console.log('receive electron msg:', msg);
					if (msg === "no") {
						reminder.remindLater();
					}
				});

				child.send("ping");
			}
		}),

		vscode.commands.registerCommand('kk.electron.openUri', openUriCommand),

		vscode.commands.registerCommand('kk.demo.showWebview', showWebview),

		vscode.commands.registerCommand('kk.seat.switch_enable', () => {
			let config = vscode.workspace.getConfiguration("seat");
			let enable = config.get<boolean>('enable', true);

			if (enable) {
				config.update('enable', false);
				statusItemEnable.text = reminderOffText;
				statusItemMinutes.hide();
				vscode.window.showInformationMessage('seat reminder off');

				reminder.enable = false;
			}
			else {
				config.update('enable', true);
				statusItemEnable.text = reminderOnText;
				statusItemMinutes.show();
				vscode.window.showInformationMessage('seat reminder on');

				reminder.enable = true;
				reminder.minutespass = 0;
			}
		}),

		vscode.commands.registerCommand('kk.seat.set_minutes', () => {
			let config = vscode.workspace.getConfiguration("seat");

			let items: vscode.QuickPickItem[] = [
				{ label: "1", description: "每[1]分钟提醒一次" },
				{ label: "30", description: "每[30]分钟提醒一次" },
				{ label: "45", description: "每[45]分钟提醒一次" },
				{ label: "60", description: "每[60]分钟提醒一次" },
				{ label: "90", description: "每[90]分钟提醒一次" },
				{ label: "120", description: "每[120]分钟提醒一次" }
			];

			vscode.window.showQuickPick(items).then(selection => {
				if (selection) {
					statusItemMinutes.text = formatMinutesText(0, selection.label);
					config.update('reminderIntervalInMinutes', parseInt(selection.label));

					reminder.minutes = parseInt(selection.label);
					reminder.minutespass = 0;
				}
			});
		})
	];

	context.subscriptions.push(...disposables);

	vscode.workspace.onDidChangeTextDocument((ev) => {
		console.log(ev);
		if (reminder) {
			reminder.active();
		}
	});
}

// this method is called when your extension is deactivated
export function deactivate() { }


function openUriCommand(): void {

	vscode.window.showInputBox({
		value: "",
		placeHolder: "https://www.google.com",
		prompt: "insert website to load",
		validateInput: (value) => value ? "" : "Enter a valid uri"
	}).then(async (value) => {

		if (value) {
			const electron_app = path.resolve(__dirname, `electron-main.js`);

			const params = {
				version: 'v12.0.2',
				params: [value]
			};
			const process = await vscode.commands.executeCommand<ChildProcess>('seat.erun', electron_app, params);

			if (process) {
				process.on("message", (message: any) => {
					console.log(message);
				}
				);
			} else {
				console.log("no process ...");
			}
		}
	});
}

interface TestDialogResult {
	name: string;
}

async function showWebview(): Promise<void> {
	// const testDir = path.resolve(__dirname, '../src/dlg');
	const testDir = path.resolve(__dirname, './kedou');

	const d = new dialog.WebviewDialog<TestDialogResult>('webview-dialog-test', testDir, 'index.html');

	const result: TestDialogResult | null = await d.getResult();

	if (result) {
		vscode.window.showInformationMessage(
			"Webview dialog result: " + JSON.stringify(result));
	} else {
		vscode.window.showInformationMessage(
			"The webview dialog was cancelled.");
	}
}

